src/oidc_child/oidc_child_curl.c

```diff
@@ -0,0 +1,462 @@
+        res = curl_easy_setopt(curl_ctx, CURLOPT_POSTFIELDS, post_data);
+        if (res != CURLE_OK) {
+            DEBUG(SSSDBG_OP_FAILURE, "Failed to add data to POST request.\n");
+            ret = EIO;
+            goto done;
+        }
+    }
+
+    if (token != NULL) {
+        res = curl_easy_setopt(curl_ctx, CURLOPT_HTTPAUTH, CURLAUTH_BEARER);
+        if (res != CURLE_OK) {
+            DEBUG(SSSDBG_OP_FAILURE, "Failed to set HTTP auth.\n");
+            ret = EIO;
+            goto done;
+        }
+        res = curl_easy_setopt(curl_ctx, CURLOPT_XOAUTH2_BEARER, token);
+        if (res != CURLE_OK) {
+            DEBUG(SSSDBG_OP_FAILURE, "Failed to add token.\n");
+            ret = EIO;
+            goto done;
+        }
+    }
+
+    ret = EOK;
+done:
+
+    return ret;
+}
+
+static errno_t do_http_request(struct devicecode_ctx *dc_ctx, const char *uri,
+                               const char *post_data, const char *token)
+{
+    CURL *curl_ctx = NULL;
+    CURLcode res;
+    int ret;
+    long resp_code;
+    struct curl_slist *headers = NULL;
+
+    headers = curl_slist_append(headers, "Accept: application/json");
+
+    curl_ctx = curl_easy_init();
+    if (curl_ctx == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to initialize curl.\n");
+        ret = EIO;
+        goto done;
+    }
+
+    ret = set_http_opts(curl_ctx, dc_ctx, uri, post_data, token, headers);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to set http options.\n");
+        goto done;
+    }
+
+    res = curl_easy_perform(curl_ctx);
+    if (res != CURLE_OK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to send request.\n");
+        ret = EIO;
+        goto done;
+    }
+
+    res = curl_easy_getinfo(curl_ctx, CURLINFO_RESPONSE_CODE, &resp_code);
+    if (res != CURLE_OK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to get response code.\n");
+        ret = EIO;
+        goto done;
+    }
+
+    if (resp_code != 200) {
+        DEBUG(SSSDBG_OP_FAILURE, "Request failed, response code is [%ld].\n",
+                                 resp_code);
+        ret = EIO;
+        goto done;
+    }
+
+    ret = EOK;
+done:
+    curl_slist_free_all(headers);
+    curl_easy_cleanup(curl_ctx);
+    return ret;
+}
+
+#define AZURE_EXPECT_CODE "The request body must contain the following parameter: 'code'."
+
+errno_t get_token(TALLOC_CTX *mem_ctx,
+                  struct devicecode_ctx *dc_ctx, const char *client_id,
+                  const char *client_secret,
+                  bool get_device_code)
+{
+    CURL *curl_ctx = NULL;
+    CURLcode res;
+    int ret;
+    size_t waiting_time = 0;
+    char *error_description = NULL;
+    char *post_data = NULL;
+    const char *post_data_tmpl = "grant_type=urn:ietf:params:oauth:grant-type:device_code&client_id=%s&%s=%s";
+    struct curl_slist *headers = NULL;
+    bool azure_fallback = false;
+
+    headers = curl_slist_append(headers, "Accept: application/json");
```

Comment:
"Accept: application/json" string is copy&pasted.

Would it be possible to re-use `do_http_request()` in general?

Reply:
`do_http_request()` always starts a new request and here I would like to keep the connection (especially the ssl tunnel) open and just send the request in regular interval until it succeeds.

Reply:
Ok, but I still think having a const/define for a string literal "Accept: application/json" makes sense.

Reply:
(But I don't insist on addressing this comment.)

Reply:
fixed